lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp

```diff
@@ -0,0 +1,261 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _SEEN
+#define DEREGISTRATION_REQUEST_SEEN
+
+#include <sstream>
+#include <thread>
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include "log.h"
+#include "intertask_interface_types.h"
+#include "intertask_interface.h"
+#include "directoryd.h"
+#include "conversions.h"
+#include "bstrlib.h"
+#ifdef __cplusplus
+};
+#endif
+#include "amf_app_ue_context_and_proc.h"
+#include "ngap_messages_types.h"
+
+using namespace std;
+namespace magma5g {
+extern task_zmq_ctx_t amf_app_task_zmq_ctx;
+/*
+ * AMBR calculation based on 9.11.4.14 of 24-501
+ */
+void ambr_calculation_pdu_session(
+    smf_context_t* smf_context, uint64_t* dl_pdu_ambr, uint64_t* ul_pdu_ambr) {
+  if ((smf_context->dl_ambr_unit == 0) || (smf_context->ul_ambr_unit == 0) ||
+      (smf_context->dl_session_ambr == 0) ||
+      (smf_context->dl_session_ambr == 0)) {
+    /* AMBR has not been populated till now and default assigned
+     * TODO rechaeck the default values 64kbps & 32mbps
+     */
+    *dl_pdu_ambr = (64 * 32768);
+    *ul_pdu_ambr = (64 * 32768);
+  } else {
+    // refer 24-501 9.11.4.14
+    switch (smf_context->dl_ambr_unit) {
+      case 1:
+        *dl_pdu_ambr = (1 * smf_context->dl_session_ambr);
+        break;
+      case 2:
+        *dl_pdu_ambr = (4 * smf_context->dl_session_ambr);
+        break;
+      case 3:
+        *dl_pdu_ambr = (16 * smf_context->dl_session_ambr);
+        break;
+      case 4:
+        *dl_pdu_ambr = (64 * smf_context->dl_session_ambr);
+        break;
+        // many more to be coded
+      default:
+        *dl_pdu_ambr = (256 * smf_context->dl_session_ambr);
+        break;
+    }
+
+    switch (smf_context->ul_ambr_unit) {
+      case 1:
+        *ul_pdu_ambr = (1 * smf_context->ul_session_ambr);
+        break;
+      case 2:
+        *ul_pdu_ambr = (4 * smf_context->ul_session_ambr);
+        break;
+      case 3:
+        *ul_pdu_ambr = (16 * smf_context->ul_session_ambr);
+        break;
+      case 4:
+        *ul_pdu_ambr = (64 * smf_context->ul_session_ambr);
+        break;
+        // many more to be coded
+      default:
+        *ul_pdu_ambr = (256 * smf_context->ul_session_ambr);
+        break;
+    }
+  }
+}
+
+/*
+ * As per Baicel capture and message flow, the function to be called
+ * before sending gRPC message to SMF with avilable information in
+ * smf_context or default.
+ * No NAS message is involved and direct itti_message is sent to NGAP.
+ * The transfer message is bstring and heavily nested structure is
+ * converted to bstring before sending to NGAP
+ */
+int pdu_session_resource_setup_request(
+    ue_m5gmm_context_s* ue_context, amf_ue_ngap_id_t amf_ue_ngap_id) {
+  if (!ue_context) {
+    // TODO
```

Comment:
you cannot leave this as is. log an error message and return error code. If ue_context is nullptr, this will lead to segfault on line 108.

Reply:
taken care

Reply:
what is to be done further here since you still have the `// TODO` comment, I expect more explanation about what needs to be done and incomplete.